pyro/contrib/funsor/handlers/replay_messenger.py

```diff
@@ -13,7 +13,11 @@ class ReplayMessenger(OrigReplayMessenger):
     """
     def _pyro_sample(self, msg):
         name = msg["name"]
-        if self.trace is not None and name in self.trace:
+        msg["replay_active"] = True  # indicate replaying so importance weights can be scaled
```

Comment:
Can you explain the `replay_*` annotations to me? Is there a deep reason these are required in `contrib.funsor` but not in main Pyro? Or asked differently, where does the corresponding logic live in main Pyro? Is it spread between `Distribution.score_parts` and `Trace_ELBO.loss_and_grads()`?

Reply:
I added these to support `poutine.scale` and subsampling in `TraceEnum_ELBO` specifically, while also attempting to make `terms_from_trace` a pure function of a single trace that I could factor out of the ELBO computation.  They are only used within `terms_from_trace`, and are basically small hacks.

The corresponding logic for scaling in the original `TraceEnum_ELBO`, which I condensed but stuck to fairly closely, is spread across the various helpers in `pyro/infer/traceenum_elbo.py`, especially `compute_model_factors`.

The use of `replay_active` is in identifying *model* sites that depend on *model-enumerated* variables, in particular, distinguishing them from enumerated guide sites, since otherwise `terms_from_trace` does not distinguish between model and guide traces.  Within `TraceEnum_ELBO`, log-probabilities at these sites must be scaled for subsampling after all model-enumerated variables are contracted, rather than immediately.  `replay_active` could be replaced by an argument to `terms_from_trace`.

The use of `replay_skipped` is in identifying *model* sites that were enumerated or sampled in the *model*.  Since there is no corresponding guide site, the unnormalized importance weight for such a site is 1, and it would be incorrect to include a `log_prob` term in the model's `log_factors`.  

`replay_skipped` is an optimization. It could be removed by setting `node["funsor"]["log_prob"] = node["funsor"]["log_prob"] * 0.` or `node["funsor"]["log_prob"] = to_funsor(0.)` in `ReplayMessenger._pyro_post_sample`, but the former would be computationally wasteful without https://github.com/pyro-ppl/funsor/issues/235 and the latter would break some of the other tests in `contrib.funsor` that currently rely on checking the shapes of every `log_prob` tensor.

Reply:
Thanks for explaining!